### PR TITLE
Notification dismissal state tracking 

### PR DESCRIPTION
**Pull request by samgst-amazon**
The user will be provided with an option to Dismiss the message. Once a user has clicked on Dismiss in any of the UI components, they will not be shown the same notification again. 

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)

## Description

To keep a track of which notifications are not shown again, notification ids of these messages will be stored to the aws.xml file(persistent storage). This set is retrieved while processing notifications and used to filter out previously dismissed notifications from being further processed and shown. 

## Checklist
- [ ] My code follows the code style of this project
- [ ] I have added tests to cover my changes
- [ ] A short description of the change has been added to the **[CHANGELOG](https://github.com/aws/aws-toolkit-jetbrains/blob/master/CONTRIBUTING.md#contributing-via-pull-requests)** if the change is customer-facing in the IDE.
- [ ] I have added metrics for my changes (if required)
 
## License
I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: plugins/core/jetbrains-community/src/software/aws/toolkits/jetbrains/core/notifications/ProcessNotificationsBase.kt [hunk 1]
```diff
@@ -3,27 +3,59 @@
...
+
+        notifications?.let { notificationsList ->
+            val (startupNotifications, emergencyNotifications) = notificationsList.notifications
+                ?.partition { notification ->
+                    notification.schedule.type.equals("StartUp", ignoreCase = true)
+                }
+                ?: Pair(emptyList(), emptyList())
+
+            val dismissalState = NotificationDismissalState.getInstance()
+            (if (isStartup) startupNotifications else emptyList())
```
Comment (manodnyab) [inline plugins/core/jetbrains-community/src/software/aws/toolkits/jetbrains/core/notifications/ProcessNotificationsBase.kt:54]: Should this part handle the fact that startup notifications should be shown only once per session?

Comment (manodnyab) [reply plugins/core/jetbrains-community/src/software/aws/toolkits/jetbrains/core/notifications/ProcessNotificationsBase.kt:54]: nit: can we move the handling into separate list handling for readability

Comment (samgst-amazon) [reply plugins/core/jetbrains-community/src/software/aws/toolkits/jetbrains/core/notifications/ProcessNotificationsBase.kt:54]: yes and yes, I was passing the boolean in from the polling service but I think the flag can live here instead.

**Review by manodnyab — APPROVED** [review]

File: plugins/core/jetbrains-community/src/software/aws/toolkits/jetbrains/utils/NotificationUtils.kt [hunk 2]
```diff
@@ -67,7 +68,7 @@ fun notifyStickyWithData(
             object : AnAction("Dismiss") {
                 override fun actionPerformed(e: AnActionEvent) {
                     ProcessNotificationsBase.showBannerNotification.remove(id)
```
Comment (rli) [inline plugins/core/jetbrains-community/src/software/aws/toolkits/jetbrains/utils/NotificationUtils.kt:70]: state should not be stored in companion objects

File: plugins/core/jetbrains-community/src/software/aws/toolkits/jetbrains/core/notifications/ProcessNotificationsBase.kt [hunk 3]
```diff
@@ -15,16 +15,20 @@ import com.intellij.openapi.project.Project
...
     val mapper = jacksonObjectMapper().configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false)
 }
 
 @Service(Service.Level.PROJECT)
-class ProcessNotificationsBase {
+class ProcessNotificationsBase(
+    private val project: Project,
+) {
     private val notifListener = mutableListOf<NotifListener>()
+    private var isStartup: AtomicBoolean = AtomicBoolean(true)
```
Comment (rli) [inline plugins/core/jetbrains-community/src/software/aws/toolkits/jetbrains/core/notifications/ProcessNotificationsBase.kt:29]: wouldn't you want startup notifications to propagate consistently across all projects on startup?

File: plugins/core/jetbrains-community/src/software/aws/toolkits/jetbrains/core/notifications/NotificationServiceInitializer.kt [hunk 4]
```diff
@@ -14,7 +14,7 @@ internal class NotificationServiceInitializer : ProjectActivity {
     override suspend fun execute(project: Project) {
         if (initialized.compareAndSet(false, true)) {
             val service = NotificationPollingService.getInstance()
-            ProcessNotificationsBase()
+            ProcessNotificationsBase(project)
```
Comment (rli) [inline plugins/core/jetbrains-community/src/software/aws/toolkits/jetbrains/core/notifications/NotificationServiceInitializer.kt:17]: this does not initialize the project service correctly

File: plugins/core/jetbrains-community/src/software/aws/toolkits/jetbrains/core/notifications/NotificationPersistantState.kt
Comment (rli) [inline plugins/core/jetbrains-community/src/software/aws/toolkits/jetbrains/core/notifications/NotificationPersistantState.kt:1]: tests?

File: plugins/core/jetbrains-community/src/software/aws/toolkits/jetbrains/core/notifications/NotificationPersistantState.kt [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+
+    var etag: String?
+        get() = state.etag
+        set(value) {
+            state.etag = value
+        }
+
+    companion object {
+        fun getInstance(): NotificationEtagState =
+            ApplicationManager.getApplication().getService(NotificationEtagState::class.java)
```
Comment (rli) [inline plugins/core/jetbrains-community/src/software/aws/toolkits/jetbrains/core/notifications/NotificationPersistantState.kt:56]:
```suggestion
        fun getInstance(): NotificationEtagState =
            ApplicationManager.getApplication().getService<NotificationEtagState>()
```

File: plugins/core/jetbrains-community/src/software/aws/toolkits/jetbrains/core/notifications/NotificationPersistantState.kt [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+    fun isDismissed(notificationId: String): Boolean =
+        state.dismissedNotificationIds.contains(notificationId)
+
+    fun dismissNotification(notificationId: String) {
+        state.dismissedNotificationIds.add(notificationId)
+    }
+
+    companion object {
+        fun getInstance(): NotificationDismissalState =
+            ApplicationManager.getApplication().getService(NotificationDismissalState::class.java)
```
Comment (rli) [inline plugins/core/jetbrains-community/src/software/aws/toolkits/jetbrains/core/notifications/NotificationPersistantState.kt:30]:
```suggestion
        fun getInstance(): NotificationDismissalState =
            ApplicationManager.getApplication().getService<NotificationDismissalState>()
```

File: plugins/core/jetbrains-community/src/software/aws/toolkits/jetbrains/core/notifications/ProcessNotificationsBase.kt [hunk 7]
```diff
@@ -39,8 +43,27 @@ class ProcessNotificationsBase {
...
 
     fun retrieveStartupAndEmergencyNotifications() {
-        // TODO: separates notifications into startup and emergency
-        // iterates through the 2 lists and processes each notification(if it isn't dismissed)
+        val notifications = getNotificationsFromFile()
+
+        notifications?.let { notificationsList ->
+            val (startupNotifications, emergencyNotifications) = notificationsList.notifications
+                ?.partition { notification ->
+                    notification.schedule.type.equals("StartUp", ignoreCase = true)
```
Comment (rli) [inline plugins/core/jetbrains-community/src/software/aws/toolkits/jetbrains/core/notifications/ProcessNotificationsBase.kt:51]: should be modeled in the deserializer

File: plugins/core/src/main/resources/META-INF/plugin.xml [hunk 8]
```diff
@@ -21,6 +21,7 @@
     <extensions defaultExtensionNs="com.intellij">
         <!-- each plugin needs its own instance of these -->
         <applicationService serviceImplementation="software.aws.toolkits.jetbrains.core.notifications.NotificationEtagState"/>
+        <applicationService serviceImplementation="software.aws.toolkits.jetbrains.core.notifications.NotificationDismissalState"/>
```
Comment (rli) [inline plugins/core/src/main/resources/META-INF/plugin.xml:24]: these can be light services

File: plugins/core/jetbrains-community/src/software/aws/toolkits/jetbrains/core/notifications/NotificationPersistantState.kt [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+package software.aws.toolkits.jetbrains.core.notifications
+
+import com.intellij.openapi.application.ApplicationManager
+import com.intellij.openapi.components.PersistentStateComponent
+import com.intellij.openapi.components.State
+import com.intellij.openapi.components.Storage
+
+@State(name = "notificationDismissals", storages = [Storage("aws.xml")])
+class NotificationDismissalState : PersistentStateComponent<NotificationDismissalConfiguration> {
+    private var state = NotificationDismissalConfiguration()
```
Comment (rli) [inline plugins/core/jetbrains-community/src/software/aws/toolkits/jetbrains/core/notifications/NotificationPersistantState.kt:13]: val

File: plugins/core/jetbrains-community/src/software/aws/toolkits/jetbrains/core/notifications/NotificationPersistantState.kt [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+    }
+}
+
+data class NotificationDismissalConfiguration(
+    var dismissedNotificationIds: MutableSet<String> = mutableSetOf(),
+)
+
+@State(name = "notificationEtag", storages = [Storage("aws.xml")])
+class NotificationEtagState : PersistentStateComponent<NotificationEtagConfiguration> {
+    private var state = NotificationEtagConfiguration()
```
Comment (rli) [inline plugins/core/jetbrains-community/src/software/aws/toolkits/jetbrains/core/notifications/NotificationPersistantState.kt:40]: val

File: plugins/core/jetbrains-community/src/software/aws/toolkits/jetbrains/core/notifications/NotificationPersistantState.kt [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+// SPDX-License-Identifier: Apache-2.0
+
+package software.aws.toolkits.jetbrains.core.notifications
+
+import com.intellij.openapi.application.ApplicationManager
+import com.intellij.openapi.components.PersistentStateComponent
+import com.intellij.openapi.components.State
+import com.intellij.openapi.components.Storage
+
+@State(name = "notificationDismissals", storages = [Storage("aws.xml")])
```
Comment (rli) [inline plugins/core/jetbrains-community/src/software/aws/toolkits/jetbrains/core/notifications/NotificationPersistantState.kt:11]: consider using `BaseState`

File: plugins/core/jetbrains-community/src/software/aws/toolkits/jetbrains/core/notifications/NotificationPersistantState.kt (hunk 5 above)
Comment (samgst-amazon) [reply plugins/core/jetbrains-community/src/software/aws/toolkits/jetbrains/core/notifications/NotificationPersistantState.kt:56]: I'm getting a syntax error when I do this. I think this is for newer versions of Intellij SDK. I don't see any other usages of the bottom code in our repository, but I do see the ::class.java version.

Comment (manodnyab) [reply plugins/core/jetbrains-community/src/software/aws/toolkits/jetbrains/core/notifications/NotificationPersistantState.kt:56]: can you try using getInstance(): NotificationETagState = service()

Comment (samgst-amazon) [reply plugins/core/jetbrains-community/src/software/aws/toolkits/jetbrains/core/notifications/NotificationPersistantState.kt:57]: that works!

File: plugins/core/jetbrains-community/src/software/aws/toolkits/jetbrains/core/notifications/NotificationCustomDeserializers.kt [hunk 12]
```diff
@@ -114,6 +114,12 @@ class NotConditionDeserializer : JsonDeserializer<NotificationExpression.NotCond
     }
 }
 
+// Create a custom deserializer if needed
+class NotificationTypeDeserializer : JsonDeserializer<NotificationScheduleType>() {
+    override fun deserialize(p: JsonParser, ctxt: DeserializationContext): NotificationScheduleType =
+        NotificationScheduleType.fromString(p.valueAsString)
+}
+
```
Comment (rli) [inline plugins/core/jetbrains-community/src/software/aws/toolkits/jetbrains/core/notifications/NotificationCustomDeserializers.kt:122]: not convinced we actually need a custom deserializer but ok

**Review by rli — APPROVED** [review]

**Pull request merged**